pyro/infer/tracegraph_elbo.py

```diff
@@ -282,28 +323,31 @@ def _loss_and_grads_particle(self, weight, model_trace, guide_trace):
         model_trace.compute_batch_log_pdf(site_filter=lambda name, site: name in model_vec_md_nodes)
         model_trace.log_pdf()
 
-        # compute elbo for reparameterized nodes
         non_reparam_nodes = set(guide_trace.nonreparam_stochastic_nodes)
+        # collect parameters to train from model and guide
+        trainable_params, trainable_params_minus_bl_params = _get_sorted_params(model_trace,
+                                                                                guide_trace, non_reparam_nodes)
+
+        # compute elbo for reparameterized nodes
         elbo, surrogate_elbo = _compute_elbo_reparam(model_trace, guide_trace, non_reparam_nodes)
 
         # the following computations are only necessary if we have non-reparameterizable nodes
         baseline_loss = 0.0
         if non_reparam_nodes:
             downstream_costs = _compute_downstream_costs(
                     model_trace, guide_trace,  model_vec_md_nodes, guide_vec_md_nodes, non_reparam_nodes)
-            surrogate_elbo_term, baseline_loss = _compute_elbo_non_reparam(
-                    guide_trace, guide_vec_md_nodes, non_reparam_nodes, downstream_costs)
-            surrogate_elbo += surrogate_elbo_term
-
-        # collect parameters to train from model and guide
-        trainable_params = set(site["value"]
-                               for trace in (model_trace, guide_trace)
-                               for site in trace.nodes.values()
-                               if site["type"] == "param")
+            baseline_loss = _compute_elbo_non_reparam(
+                    guide_trace, guide_vec_md_nodes, non_reparam_nodes, downstream_costs,
+                    trainable_params_minus_bl_params)
 
         if trainable_params:
             surrogate_loss = -surrogate_elbo
-            torch_backward(weight * (surrogate_loss + baseline_loss))
+            # XXX this try block is a hack that will go away when we phase out backward()
```

Comment:
What is the `RuntimeError` that is raised? Can we move this try block into `torch_backward`?

Reply:
it has to do with the fact that backward doesn't have an allow_unused=True flag
(so if there is nothing to differentiate, it fails)

Reply:
Could you add that to the comment in this code?

Reply:
added more comments